Client/Frontend/Browser/HomePanel/NTPDataSource.swift

```diff
@@ -0,0 +1,240 @@
+                // Trimming to fixed length to release older backgrounds
+                
+                self.lastBackgroundChoices = self.lastBackgroundChoices
+                    .suffix(min(backgroundSet.count - 1, NTPDataSource.numberOfDuplicateAvoidance))
+                return chosenIndex
+            }
+        }()
+        
+        // Force back to `0` if at end
+        backgroundRotationCounter %= NTPDataSource.sponsorshipShowRate
+        // Increment regardless, this is a counter, not an index, so smallest should be `1`
+        backgroundRotationCounter += 1
+        
+        return (backgroundSet[backgroundIndex], backgroundType)
+    }
+    
+    private func loadData(file: String) -> Data? {
+        guard let filePath = Bundle.main.path(forResource: file, ofType: "json") else {
+            return nil
+        }
+        
+        do {
+            let backgroundData = try Data(contentsOf: URL(fileURLWithPath: filePath))
+            return backgroundData
+        } catch {
+            Logger.browserLogger.error("Failed to get bundle path for \(file)")
+        }
+        
+        return nil
+    }
+}
+
+extension NTPDataSource: NTPDownloaderDelegate {
+    func onSponsorUpdated(sponsor: NTPSponsor?) {
+        self.sponsor = sponsor
+        // Force to set up basic favorites if it hasn't been done already.
+        initializeFavorites?(nil)
+    }
+    
+    func onThemeUpdated(theme: CustomTheme?) {
+        self.customTheme = theme
+        
+        if Preferences.NewTabPage.preloadedFavoritiesInitialized.value {
+            replaceFavoritesIfNeeded?(theme?.topSites)
+        } else {
+            initializeFavorites?(theme?.topSites)
+        }
+        
+    }
+}
+
+extension NTPDataSource: PreferencesObserver {
+    func preferencesDidChange(for key: String) {
+        let sponsoredPref = Preferences.NewTabPage.backgroundSponsoredImages
+        let customThemePref = Preferences.NewTabPage.selectedCustomTheme
+        let installedThemesPref = Preferences.NewTabPage.installedCustomThemes
+        
+        switch key {
+        case sponsoredPref.key:
+            if sponsoredPref.value {
+                // Issue download only when no custom theme is currently set.
+                if customThemePref.value == nil {
+                    Preferences.NTP.ntpCheckDate.value = nil
+                    downloader.delegate = self
+                    fetchSpecificResource(.sponsor)
+                }
+
+            } else {
+                downloader.delegate = nil
+                do {
+                    try downloader.removeCampaign(type: .sponsor)
+                } catch {
+                    Logger.browserLogger.error(error)
+                }
+            }
+        case customThemePref.key:
+            downloader.delegate = self
+            
+            let installedThemes = installedThemesPref.value
+            if let theme = customThemePref.value, !installedThemes.contains(theme) {
```

Comment:
Is there a cleanup done after user disables the custom pref, may be after a couple of days after the pref changed?

Reply:
No, the user must be able to switch between themes until the campaign is ended

Only way to remove assets right now is when the campaign is ended on server side